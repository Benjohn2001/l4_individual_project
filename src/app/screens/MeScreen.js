import React, { useEffect, useState } from "react";
import {
  View,
  Image,
  TextInput,
  Text,
  SafeAreaView,
  Alert,
  ActivityIndicator,
} from "react-native";
import {
  getDatabase,
  ref,
  onValue,
  query,
  get,
  update,
} from "firebase/database";
import Modal from "react-native-modal";
import * as ImagePicker from "expo-image-picker";
import {
  getStorage,
  getDownloadURL,
  uploadBytes,
  ref as stref,
} from "firebase/storage";
import AppButtonPurple from "../components/AppButtonPurple";
import COLOURS from "../assets/colours";
import ProfileBox from "../components/ProfileBox";
import RowItem from "../components/RowItem";
import PillButton from "../components/PillButton";
import { auth } from "../../firebase";
import TwoButtonStack from "../components/TwoButtonStack";

function MeScreen() {
  const db = getDatabase();
  const user = auth.currentUser;
  const { uid } = user;
  const unamesRef = query(ref(db, `/users/${uid}`));

  const [nameFull, setNameFull] = useState("");
  const [userName, setUserName] = useState("");
  const [bio, setBio] = useState("");
  const [bioNew, setBioNew] = useState("");
  const [pic, setPic] = useState("");
  const [isBioModalVisible, setBioModalVisible] = React.useState(false);
  const [userNameNew, setuserNameNew] = useState("");
  const [isuserNameModalVisible, setuserNameModalVisible] =
    React.useState(false);
  const [isPicModalVisible, setPicModalVisible] = React.useState(false);
  const [picUp, setPicUp] = React.useState("");
  const [uploading, setUploading] = React.useState(false);

  const pickImage = async () => {
    const result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.All,
      allowsEditing: true,
      aspect: [4, 4],
      quality: 0.2,
    });

    if (!result.cancelled) {
      setPicUp(result.uri);
    }
  };

  async function uploadImageAsync(uri) {
    const blob = await new Promise((resolve, reject) => {
      const xhr = new XMLHttpRequest();
      xhr.onload = function load() {
        resolve(xhr.response);
      };
      xhr.onerror = function error() {
        reject(new TypeError("Network request failed"));
      };
      xhr.responseType = "blob";
      xhr.open("GET", uri, true);
      xhr.send(null);
    });

    const fileRef = stref(getStorage(), `images/profilePics/${uid}`);
    await uploadBytes(fileRef, blob);

    blob.close();

    return getDownloadURL(fileRef);
  }

  useEffect(() => {
    onValue(
      unamesRef,
      (snapshot) => {
        if (snapshot.exists) {
          const data = snapshot.val();
          setNameFull(`${data.firstName} ${data.lastName}`);
          setUserName(data.userName);
          setBio(data.bio);
          if (data.profilePic !== undefined) {
            getDownloadURL(stref(getStorage(), data.profilePic)).then((url) => {
              setPic(url);
            });
          } else {
            setPic(
              Image.resolveAssetSource(
                require("../assets/defaultProfilePic.png")
              ).uri
            );
          }
        }
      },
      {}
    );
  }, []);

  return (
    <SafeAreaView className="flex-1 bg-primaryPurple">
      {pic ? (
        <View className="items-center pt-14">
          <Image className="w-44 h-44 rounded-full" source={{ uri: pic }} />
          <View className="py-3">
            <ProfileBox name={nameFull} username={userName} bio={bio} />
          </View>
          <RowItem
            title="Change my username"
            icon="edit-3"
            onPress={() => {
              setuserNameModalVisible(true);
            }}
            color="black"
          />
          <RowItem
            title="Change my profile picture"
            icon="camera"
            onPress={() => {
              setPicModalVisible(true);
            }}
            color="black"
          />
          <RowItem
            title="Change my bio"
            icon="edit-3"
            onPress={() => {
              setBioModalVisible(true);
            }}
            color="black"
          />
        </View>
      ) : (
        <View className="justify-center items-center flex-1">
          <ActivityIndicator size="large" color="#6B4EFF" />
          <Text className="text-center">Loading Profile</Text>
        </View>
      )}
      <View className="mt-auto py-3">
        <PillButton
          title="Share my profile"
          onPress={() => {
            alert("Share my profile");
          }}
          icon="share-2"
        />
      </View>

      <Modal
        isVisible={isBioModalVisible}
        onBackdropPress={() => setBioModalVisible(false)}
        className="items-center"
      >
        <View className=" w-11/12 h-50 items-center py-3 px-3 bg-secondaryPurple rounded-2xl">
          <Text className="text-2xl font-bold  pb-2">Change my bio</Text>
          <Text className="text-center text-gray-500 py-4">
            Enter a new bio {"\n"} Max 50 characters
          </Text>
          <TextInput
            className="bg-white my-5 w-full mx-3 h-12 rounded-md"
            placeholder="bio"
            underlineColorAndroid="transparent"
            cursorColor={COLOURS.darkerPurple}
            maxLength={50}
            value={bioNew}
            onChangeText={(val) => setBioNew(val)}
          />
          <AppButtonPurple
            title="Change Bio"
            onPress={async () => {
              await update(ref(db, `/users/${uid}`), {
                bio: bioNew,
              }).then(() => {
                Alert.alert("Success", "Bio changed succesfully");
                setBioModalVisible(false);
              });
              setBioNew("");
            }}
          />
        </View>
      </Modal>
      <Modal
        isVisible={isuserNameModalVisible}
        onBackdropPress={() => setuserNameModalVisible(false)}
        className="items-center"
      >
        <View className=" w-11/12 h-50 items-center py-3 px-3 bg-secondaryPurple rounded-2xl">
          <Text className="text-2xl font-bold  pb-2">Change my username</Text>
          <Text className="text-center text-gray-500 py-4">
            Enter a new username
          </Text>
          <TextInput
            className="bg-white my-5 w-full mx-3 h-12 rounded-md"
            placeholder="username"
            underlineColorAndroid="transparent"
            cursorColor={COLOURS.darkerPurple}
            maxLength={15}
            value={userNameNew}
            onChangeText={(val) => setuserNameNew(val)}
          />
          <AppButtonPurple
            title="Change Username"
            onPress={async () => {
              const snapshot = await get(unamesRef);
              if (snapshot.val() !== null) {
                Alert.alert("Username taken", "Please select a new username");
              } else {
                await update(ref(db, `/users/${uid}`), {
                  userName: userNameNew,
                }).then(() => {
                  Alert.alert("Success", "Username changed succesfully");
                });
                setuserNameModalVisible(false);
              }
              setuserNameNew("");
            }}
          />
        </View>
      </Modal>
      {uploading ? (
        <View className="justify-center items-center flex-1">
          <ActivityIndicator size="large" color="#6B4EFF" />
          <Text className="text-center">Uploading</Text>
        </View>
      ) : (
        <Modal
          isVisible={isPicModalVisible}
          onBackdropPress={() => setPicModalVisible(false)}
          className="items-center"
        >
          <View className=" w-11/12 h-50 items-center py-3 px-3 bg-secondaryPurple rounded-2xl">
            <Text className="text-2xl font-bold  pb-2">Change profile pic</Text>
            <Text className="text-center text-gray-500 py-4">
              Select a new profile pic
            </Text>
            <View className="pb-5">
              {picUp && (
                <Image
                  source={{ uri: picUp }}
                  className="h-28 w-28 rounded-full"
                />
              )}
            </View>
            <TwoButtonStack
              title1="Upload new picture"
              onPress1={() => {
                pickImage();
              }}
              title2="Save picture"
              onPress2={async () => {
                if (picUp === "") {
                  Alert.alert("No image selected", "Please upload an image");
                } else {
                  setUploading(true);
                  setPicModalVisible(false);
                  const uploadUrl = await uploadImageAsync(picUp);
                  setUploading(false);
                  setPic(uploadUrl);
                  await update(ref(db, `/users/${uid}`), {
                    profilePic: `images/profilePics/${uid}`,
                  }).then(() => {
                    Alert.alert("Success", "Profile pic changed succesfully");
                  });
                }
                setPicUp("");
              }}
            />
          </View>
        </Modal>
      )}
    </SafeAreaView>
  );
}

export default MeScreen;
