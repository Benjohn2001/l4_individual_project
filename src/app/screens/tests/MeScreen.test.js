import renderer from "react-test-renderer";
import { render } from "@testing-library/react-native";
import { configure } from "enzyme/build";
import Adapter from "enzyme-adapter-react-16";
import MeScreen from "../MeScreen";

configure({ adapter: new Adapter() });

jest.mock("firebase/database", () => ({
  getDatabase: jest.fn().mockReturnThis(),
  push: jest.fn().mockReturnThis(),
  ref: jest.fn().mockReturnThis(),
  set: jest.fn().mockReturnThis(),
  query: jest.fn().mockReturnThis(),
  onValue: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/storage", () => ({
  getStorage: jest.fn().mockReturnThis(),
}));

const navigation = {
  navigate: jest.fn(),
  push: jest.fn(),
};

jest.mock("firebase/app", () => ({
  auth: jest.fn().mockReturnThis(),
  initializeApp: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/auth", () => ({
  auth: jest.fn().mockReturnThis(),
  getAuth: () => ({
    currentUser: jest.fn().mockReturnThis(),
    signOut: () => ({
      then: () => ({
        catch: jest.fn(),
      }),
    }),
  }),
}));

test("Check MeScreen loading correctly", () => {
  render(<MeScreen navigation={navigation} />);

  const domTree = renderer.create(<MeScreen />).toJSON();
  expect(domTree).toMatchSnapshot();
});

// cant fin d row items as pic not changed state
// test("Check MeScreen change uname correctly", () => {
//     const meUname = shallow(<MeScreen navigation={navigation}/>);

//     meUname.find(RowItem).at(0).simulate('press')
//     // fireEvent.press(getByText("Change my username"))
//     // fireEvent.changeText(getByPlaceholderText("username"),"123")
//     // fireEvent.press(getByText("Change Username"))
// });
