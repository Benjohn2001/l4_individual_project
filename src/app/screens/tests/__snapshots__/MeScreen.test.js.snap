// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check MeScreen loading correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      Object {
        "backgroundColor": "#C6C4FF",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "justifyContent": "center",
        },
        Object {
          "alignItems": "center",
        },
        Object {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <ActivityIndicator
      color="#6B4EFF"
      size="large"
    />
    <Text
      style={
        Array [
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      Loading Profile
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "marginTop": "auto",
        },
        Object {
          "paddingBottom": 12,
          "paddingTop": 12,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
          },
          Object {
            "paddingBottom": 12,
            "paddingTop": 12,
          },
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#6B4EFF",
            "borderBottomLeftRadius": 9999,
            "borderBottomRightRadius": 9999,
            "borderTopLeftRadius": 9999,
            "borderTopRightRadius": 9999,
            "height": 32,
            "justifyContent": "space-between",
            "marginLeft": 6,
            "marginRight": 6,
            "opacity": 1,
            "width": 176,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Object {
                "justifyContent": "center",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "paddingTop": 8,
                },
              ]
            }
          >
            <Text />
          </View>
          <View
            style={
              Array [
                Object {
                  "paddingTop": 4,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#fff",
                  },
                  Object {
                    "fontSize": 14,
                    "lineHeight": 20,
                  },
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "marginLeft": 20,
                  },
                ]
              }
            >
              Share my profile
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    transparent={true}
    visible={false}
  />
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    transparent={true}
    visible={false}
  />
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    transparent={true}
    visible={false}
  />
</RCTSafeAreaView>
`;
